javascript/ql/src/semmle/javascript/DefensiveProgramming.qll

```diff
@@ -0,0 +1,412 @@
+  private predicate isNullOrUndefined(InferredType t) {
+    t = TTNull() or
+    t = TTUndefined()
+  }
+
+  /**
+   * Holds if `t` is not `null` or `undefined`.
+   */
+  private predicate isNotNullOrUndefined(InferredType t) {
+    not isNullOrUndefined(t)
+  }
+
+  /**
+   * A value comparison for `null` and `undefined`.
+   */
+  private class NullUndefinedComparison extends UndefinedNullTest {
+
+    Expr operand;
+
+    InferredType op2type;
+
+    NullUndefinedComparison() {
+      exists (Expr op2 |
+        hasOperands(operand, op2) |
+        op2type = TTNull() and SyntacticConstants::isNull(op2)
+        or
+        op2type = TTUndefined() and SyntacticConstants::isUndefined(op2)
+      )
+    }
+
+    override boolean getTheTestResult() {
+      result = getPolarity() and
+      (
+        if this instanceof StrictEqualityTest then
+          operand.analyze().getTheType() = op2type
+        else
+          not isNotNullOrUndefined(operand.analyze().getAType())
+      )
+      or
+      result = getPolarity().booleanNot() and
+      (
+        if this instanceof StrictEqualityTest then
+          not operand.analyze().getAType() = op2type
+        else
+          not isNullOrUndefined(operand.analyze().getAType())
+      )
+    }
+
+    override Expr getOperand() {
+      result = operand
+    }
+  }
+
+  /**
+   * An expression that throws an exception if one of its subexpressions evaluates to `null` or `undefined`.
+   */
+  private class UndefinedNullCrashUse extends Expr {
+
+    Expr target;
+
+    UndefinedNullCrashUse() {
+      this.(InvokeExpr).getCallee().getUnderlyingValue() = target
+      or
+      this.(PropAccess).getBase().getUnderlyingValue() = target
+      or
+      this.(MethodCallExpr).getReceiver().getUnderlyingValue() = target
+    }
+
+    /**
+     * Gets the subexpression that will cause an exception to be thrown if it is `null` or `undefined`.
+     */
+    Expr getVulnerableSubexpression() {
+      result = target
+    }
+
+  }
+
+  /**
+   * An expression that throws an exception if one of its subexpressions is not a `function`.
+   */
+  private class NonFunctionCallCrashUse extends Expr {
+
+    Expr target;
+
+    NonFunctionCallCrashUse() {
+      this.(InvokeExpr).getCallee().getUnderlyingValue() = target
+    }
+
+    /**
+     * Gets the subexpression that will cause an exception to be thrown if it is not a `function`.
+     */
+    Expr getVulnerableSubexpression() {
+      result = target
+    }
+
+  }
+
+  /**
+   * Gets the first expression that is guarded by `guard`.
+   */
+  private Expr getAGuardedExpr(Expr guard) {
+    exists(BinaryExpr op |
+      op.getLeftOperand() = guard and
+      (op instanceof LogAndExpr or op instanceof LogOrExpr) and
+      op.getRightOperand() = result
+    )
+    or
+    exists(IfStmt c |
+      c.getCondition() = guard |
+      result = c.getAControlledStmt().getChildExpr(0) or
+      result = c.getAControlledStmt().(BlockStmt).getStmt(0).getChildExpr(0)
```

Comment:
I don't like the use of `getChildExpr(0)` here; it makes assumptions about the order of evaluation that are hard to keep track of. Do you perhaps want to use `getFirstControlFlowNode`?

Reply:
Hmm, it is quite awkward to get the first expression since `getFirstControlFlowNode` for the case we care about (`ExprStmt`) is the node itself. See eb9c57cf. We may want a proper `getFirstExpr` at a later time.